File: core/package.py
import os
import string
import itertools

import core.sign as cs
import core.utils as cu
import core.error as ce
import core.gen_cmds as cg
import core.render_ctx as cr


def parse_pkg_flags(v):
    def it():
        for x in v.split(','):
            a, b, c = x.partition('=')

            yield a, c

    return dict(it())


def parse_pkg_name(v):
    a, b, c = v.partition('(')

    r = {
        'name': a,
    }

    if b:
        r['flags'] = parse_pkg_flags(c[:-1])

    return r


def make_selector(v, flags):
    v = parse_pkg_name(v)

    v['flags'] = cu.dict_dict_update(flags, v.get('flags', {}))

    return v


def rlist(l):
    return list(reversed(list(l)))


def visit_node(n, f, k):
    s = set()

    def v(l):
        kk = k(l)

        if kk not in s:
            s.add(kk)

            for x in rlist(f(l)):
                yield from v(x)

            yield l

    return rlist(v(n))


def visit_lst(l, f):
    def ff(n):
        return f(n) if n else l

    def kk(n):
        return n.uid if n else None

    return visit_node(None, ff, kk)[1:]


def add_kind(k, l):
    return ({'kind': k, 'p': x} for x in l)


def popf1(d, f):
    if f in d:
        d = cu.copy_dict(d)
        d.pop(f)

    return d


def popf(d, *f):
    for x in f:
        d = popf1(d, x)

    return d


ONE_LEVEL = ('setx', 'help', 'verbose', 'stage', 'rebuild', 'nostrip')


def sanitize(flags):
    return popf(flags, *ONE_LEVEL)


def parse_kind(n):
    v = n[:n.index('/')]

    return {
        'bld': 'bin',
        'etc': 'aux',
    }.get(v, v)


def fix_selector(selector, config):
    selector = cu.copy_dict(selector)

    if 'flags' not in selector:
        selector['flags'] = {}

    flags = selector['flags']

    if 'target' not in flags:
        flags['target'] = config.platform['target']

    if 'kind' not in flags:
        flags['kind'] = parse_kind(selector['name'])

    return selector


def remsuf(s, suf):
    if s.endswith(suf):
        return s[:-len(suf)]

    return s


class Package:
    def __init__(self, selector, mngr):
        self.manager = mngr
        self.selector = fix_selector(selector, self.config)
        self.pkg_name = self.calc_pkg_name()
        self.descr = cr.RenderContext(self).render()

        if self.buildable():
            self.uid = cs.UID
            self.uid = list(self.iter_build_commands())[-1]['uid']
        else:
            self.uid = cu.struct_hash(self.selector)

    @property
    def norm_name(self):
        return remsuf(remsuf(self.name, '.sh'), '/ix')

    @property
    def out_dir(self):
        return f'{self.config.store_dir}/{self.uid}-{self.pkg_name}'

    def calc_pkg_name(self):
        k = self.flags['kind']
        n = self.norm_name

        try:
            n = n[n.index('/') + 1:]
        except ValueError:
            pass

        return cu.canon_name(f'{k}-{n}')

    @property
    def uniq_id(self):
        return self.pkg_name.replace('-', '_')

    @property
    def flags(self):
        return self.selector['flags']

    @property
    def name(self):
        res = self.selector['name']

        if not res.endswith('.sh'):
            res = res + '/ix.sh'

        return res

    @property
    def config(self):
        return self.manager.config

    @property
    def target(self):
        return self.flags['target']

    @property
    def host(self):
        return self.config.platform['host']

    def host_lib_flags(self):
        return {'target': self.host, 'kind': 'lib'}

    def target_lib_flags(self):
        return sanitize(cu.dict_dict_update(self.flags, {'kind': 'lib'}))

    def calc_bin_flags(self, target):
        basef = {} if self.buildable() else self.flags

        return cu.dict_dict_update(basef, {'target': target, 'kind': 'bin'})

    def bin_flags(self):
        return self.calc_bin_flags(self.host)

    def load_package(self, n, flags):
        try:
            n['name']
        except TypeError:
            n = make_selector(n, flags)

        return self.load_package_impl(n)

    def load_package_impl(self, sel):
        # TODO(pg): proper local flags
        return self.manager.load_package(sanitize(sel), self.selector)

    def load_packages(self, l, flags):
        return (self.load_package(x, flags) for x in l)

    def bld_lib_deps(self, k):
        yield from self.descr['bld'][k]

        for p in self.bld_bin_closure():
            yield from p.descr['ind']['deps']

    def bld_host_lib_deps(self):
        return self.bld_lib_deps('host_libs')

    def bld_target_lib_deps(self):
        return self.bld_lib_deps('target_libs')

    def visit(self, lst, flags, childf):
        return visit_lst(self.load_packages(lst, flags), childf)

    @cu.cached_method
    def bld_bin_closure(self):
        return self.visit(self.descr['bld']['deps'], self.bin_flags(), lambda x: x.run_closure())

    @cu.cached_method
    def lib_closure(self):
        return self.visit(self.descr['lib']['deps'], self.target_lib_flags(), lambda x: x.lib_closure())

    @cu.cached_method
    def bld_host_lib_closure(self):
        return self.visit(self.bld_host_lib_deps(), self.host_lib_flags(), lambda x: x.lib_closure())

    @cu.cached_method
    def bld_target_lib_closure(self):
        return self.visit(self.bld_target_lib_deps(), self.target_lib_flags(), lambda x: x.lib_closure())

    def bld_lib_closure(self):
        return itertools.chain(self.bld_target_lib_closure(), self.bld_host_lib_closure())

    @cu.cached_method
    def bld_data(self):
        return self.load_data_dep(self.descr['bld']['data'])

    def iter_all_tagged_build_depends(self):
        yield from add_kind('bin', self.bld_bin_closure())
        yield from add_kind('data', self.bld_data())
        yield from add_kind('target lib', self.bld_target_lib_closure())
        yield from add_kind('host lib', self.bld_host_lib_closure())

    def iter_tagged_build_depends(self):
        for x in self.iter_all_tagged_build_depends():
            if x['p'].buildable():
                yield x

    def iter_extra_tools(self):
        if 'sem:' in str(self.descr['bld']['fetch']):
            yield 'bin/semver'

    def iter_extra_build_depends(self):
        # TODO(pg): refac
        return self.visit(self.iter_extra_tools(), self.bin_flags(), lambda x: x.run_closure())

    def iter_all_build_depends_dup(self):
        for x in self.iter_tagged_build_depends():
            yield x['p']

        yield from self.iter_extra_build_depends()

    @cu.cached_method
    def iter_all_build_depends(self):
        return list(cu.uniq_p(self.iter_all_build_depends_dup()))

    @cu.cached_method
    def iter_build_dirs(self):
        return [x.out_dir for x in self.iter_all_build_depends()]

    @cu.cached_method
    def run_deps(self):
        return list(self.load_packages(self.descr['run']['deps'], self.calc_bin_flags(self.target)))

    def load_data_dep(self, l):
        return list(self.load_packages(l, {'target': self.target, 'kind': 'aux'}))

    @cu.cached_method
    def run_data(self):
        return self.load_data_dep(self.descr['run']['data'])

    def all_run_deps(self):
        yield from self.run_deps()
        yield from self.run_data()

        # TODO(pg): think trice about it
        if self.flags['kind'] == 'lib':
            yield from self.lib_closure()

        for p in self.bld_target_lib_closure():
            yield from p.run_data()

    @cu.cached_method
    def run_closure(self):
        return visit_lst(self.all_run_deps(), lambda x: x.run_closure())

    def iter_all_runtime_depends(self):
        return filter(lambda x: x.buildable(), self.run_closure())

    def buildable(self):
        return not not self.descr['bld']['script']

    def iter_build_commands(self):
        return cg.iter_build_commands(self)

    def find_tool(self, name):
        for x in self.iter_all_build_depends():
            if x.norm_name == name:
                return x

        raise Exception(f'can not find {name} for {self.name}')

File: go/package.go
package main

import (
	"fmt"
	"strings"
)

var ONE_LEVEL = map[string]bool{
	"setx": true, "help": true, "verbose": true, "stage": true,
	"rebuild": true, "nostrip": true, "simulate_failure": true,
	"show_script": true, "skipsrc_one": true,
}

func sanitize(flags map[string]any) map[string]any {
	result := make(map[string]any, len(flags))
	for k, v := range flags {
		if !ONE_LEVEL[k] {
			result[k] = v
		}
	}
	return result
}

func mergeFlags(base, over map[string]any) map[string]any {
	result := make(map[string]any, len(base)+len(over))
	for k, v := range base {
		result[k] = v
	}
	for k, v := range over {
		result[k] = v
	}
	return result
}

type Selector struct {
	Name  string
	Flags map[string]any
}

func (s Selector) cacheKey() string {
	return structHash(map[string]any{"name": s.Name, "flags": s.Flags})
}

func remsuf(s, suf string) string {
	return strings.TrimSuffix(s, suf)
}

type Package struct {
	selector Selector
	descr    *Descr
	uid      string
	pkgName  string
	mgr      *Manager
	boot     bool

	// cached closures
	_bldBinClosure       []*Package
	_libClosure          []*Package
	_bldHostLibClosure   []*Package
	_bldTargetLibClosure []*Package
	_bldData             []*Package
	_runDeps             []*Package
	_runData             []*Package
	_runClosure          []*Package
	_allBuildDepends     []*Package
	_buildDirs           []string
	_buildCmds           []map[string]any

	_bldBinClosureDone       bool
	_libClosureDone          bool
	_bldHostLibClosureDone   bool
	_bldTargetLibClosureDone bool
	_bldDataDone             bool
	_runDepsDone             bool
	_runDataDone             bool
	_runClosureDone          bool
	_allBuildDependsDone     bool
	_buildDirsDone           bool
	_buildCmdsDone           bool
}

func (p *Package) Name() string {
	n := p.selector.Name
	if !strings.HasSuffix(n, ".sh") {
		n += "/ix.sh"
	}
	return n
}

func (p *Package) NormName() string {
	return remsuf(remsuf(p.Name(), ".sh"), "/ix")
}

func (p *Package) Flags() map[string]any {
	return p.selector.Flags
}

func (p *Package) Target() map[string]any {
	if t, ok := p.selector.Flags["target"]; ok {
		if tm, ok := t.(map[string]any); ok {
			return tm
		}
	}
	return nil
}

func (p *Package) OutDir() string {
	return fmt.Sprintf("%s/%s-%s", p.mgr.config.StoreDir(), p.uid, p.pkgName)
}

func (p *Package) Buildable() bool {
	return p.descr.Buildable()
}

func (p *Package) UniqID() string {
	return strings.ReplaceAll(p.pkgName, "-", "_")
}

func (p *Package) calcPkgName() string {
	kind := "lib"
	if k, ok := p.selector.Flags["kind"].(string); ok {
		kind = k
	}
	n := p.NormName()
	if idx := strings.IndexByte(n, '/'); idx >= 0 {
		n = n[idx+1:]
	}
	return canonName(kind + "-" + n)
}

func (p *Package) hostLibFlags() map[string]any {
	return map[string]any{"target": p.mgr.config.Host, "kind": "lib"}
}

func (p *Package) targetLibFlags() map[string]any {
	return sanitize(mergeFlags(p.Flags(), map[string]any{"kind": "lib"}))
}

func (p *Package) binFlags() map[string]any {
	return p.calcBinFlags(p.mgr.config.Host)
}

func (p *Package) calcBinFlags(target map[string]any) map[string]any {
	base := map[string]any{}
	if !p.Buildable() {
		base = p.Flags()
	}
	return mergeFlags(base, map[string]any{"target": target, "kind": "bin"})
}

func (p *Package) loadPackage(n string, flags map[string]any) *Package {
	name, pFlags := parsePkgName(n)
	merged := mergeFlags(flags, pFlags)
	sel := Selector{Name: name, Flags: merged}
	return p.mgr.LoadPackage(sanitize(sel.Flags), sel.Name)
}

func (p *Package) loadPackages(names []string, flags map[string]any) []*Package {
	var result []*Package
	for _, n := range names {
		result = append(result, p.loadPackage(n, flags))
	}
	return result
}

// visitList performs topological traversal with dedup.
func visitList(initial []*Package, childFn func(*Package) []*Package) []*Package {
	seen := map[string]bool{}
	var result []*Package

	var visit func(p *Package)
	visit = func(p *Package) {
		if seen[p.uid] {
			return
		}
		seen[p.uid] = true

		children := childFn(p)
		for i := len(children) - 1; i >= 0; i-- {
			visit(children[i])
		}
		result = append(result, p)
	}

	for i := len(initial) - 1; i >= 0; i-- {
		visit(initial[i])
	}

	// Reverse
	for i, j := 0, len(result)-1; i < j; i, j = i+1, j-1 {
		result[i], result[j] = result[j], result[i]
	}
	return result
}

func (p *Package) bldBinClosure() []*Package {
	if !p._bldBinClosureDone {
		pkgs := p.loadPackages(p.descr.Bld.Deps, p.binFlags())
		p._bldBinClosure = visitList(pkgs, func(x *Package) []*Package { return x.runClosure() })
		p._bldBinClosureDone = true
	}
	return p._bldBinClosure
}

func (p *Package) libClosure() []*Package {
	if !p._libClosureDone {
		pkgs := p.loadPackages(p.descr.Lib.Deps, p.targetLibFlags())
		p._libClosure = visitList(pkgs, func(x *Package) []*Package { return x.libClosure() })
		p._libClosureDone = true
	}
	return p._libClosure
}

func (p *Package) bldHostLibDeps() []string {
	deps := make([]string, len(p.descr.Bld.HostLibs))
	copy(deps, p.descr.Bld.HostLibs)
	for _, bp := range p.bldBinClosure() {
		deps = append(deps, bp.descr.Ind.Deps...)
	}
	return deps
}

func (p *Package) bldTargetLibDeps() []string {
	deps := make([]string, len(p.descr.Bld.TargetLibs))
	copy(deps, p.descr.Bld.TargetLibs)
	for _, bp := range p.bldBinClosure() {
		deps = append(deps, bp.descr.Ind.Deps...)
	}
	return deps
}

func (p *Package) bldHostLibClosure() []*Package {
	if !p._bldHostLibClosureDone {
		pkgs := p.loadPackages(p.bldHostLibDeps(), p.hostLibFlags())
		p._bldHostLibClosure = visitList(pkgs, func(x *Package) []*Package { return x.libClosure() })
		p._bldHostLibClosureDone = true
	}
	return p._bldHostLibClosure
}

func (p *Package) bldTargetLibClosure() []*Package {
	if !p._bldTargetLibClosureDone {
		pkgs := p.loadPackages(p.bldTargetLibDeps(), p.targetLibFlags())
		p._bldTargetLibClosure = visitList(pkgs, func(x *Package) []*Package { return x.libClosure() })
		p._bldTargetLibClosureDone = true
	}
	return p._bldTargetLibClosure
}

func (p *Package) bldData() []*Package {
	if !p._bldDataDone {
		p._bldData = p.loadPackages(p.descr.Bld.Data, map[string]any{"target": p.mgr.config.Host, "kind": "aux"})
		p._bldDataDone = true
	}
	return p._bldData
}

type taggedPkg struct {
	kind string
	pkg  *Package
}

func (p *Package) iterAllTaggedBuildDepends() []taggedPkg {
	var result []taggedPkg
	for _, x := range p.bldBinClosure() {
		result = append(result, taggedPkg{"bin", x})
	}
	for _, x := range p.bldData() {
		result = append(result, taggedPkg{"data", x})
	}
	for _, x := range p.bldTargetLibClosure() {
		result = append(result, taggedPkg{"target lib", x})
	}
	for _, x := range p.bldHostLibClosure() {
		result = append(result, taggedPkg{"host lib", x})
	}
	return result
}

func (p *Package) iterTaggedBuildDepends() []taggedPkg {
	var result []taggedPkg
	for _, tp := range p.iterAllTaggedBuildDepends() {
		if tp.pkg.Buildable() {
			result = append(result, tp)
		}
	}
	return result
}

func (p *Package) iterAllBuildDepends() []*Package {
	if !p._allBuildDependsDone {
		seen := map[string]bool{}
		var result []*Package
		for _, tp := range p.iterTaggedBuildDepends() {
			if !seen[tp.pkg.uid] {
				seen[tp.pkg.uid] = true
				result = append(result, tp.pkg)
			}
		}
		p._allBuildDepends = result
		p._allBuildDependsDone = true
	}
	return p._allBuildDepends
}

func (p *Package) iterBuildDirs() []string {
	if !p._buildDirsDone {
		for _, x := range p.iterAllBuildDepends() {
			p._buildDirs = append(p._buildDirs, x.OutDir())
		}
		p._buildDirsDone = true
	}
	return p._buildDirs
}

func (p *Package) runDeps() []*Package {
	if !p._runDepsDone {
		t := p.Target()
		if t == nil {
			t = p.mgr.config.Host
		}
		p._runDeps = p.loadPackages(p.descr.Run.Deps, p.calcBinFlags(t))
		p._runDepsDone = true
	}
	return p._runDeps
}

func (p *Package) runData() []*Package {
	if !p._runDataDone {
		p._runData = p.loadPackages(p.descr.Run.Data, map[string]any{"target": p.mgr.config.Host, "kind": "aux"})
		p._runDataDone = true
	}
	return p._runData
}

func (p *Package) allRunDeps() []*Package {
	var result []*Package
	result = append(result, p.runDeps()...)
	result = append(result, p.runData()...)

	kind, _ := p.selector.Flags["kind"].(string)
	if kind == "lib" {
		result = append(result, p.libClosure()...)
	}

	for _, tlp := range p.bldTargetLibClosure() {
		result = append(result, tlp.runData()...)
	}
	return result
}

func (p *Package) runClosure() []*Package {
	if !p._runClosureDone {
		p._runClosure = visitList(p.allRunDeps(), func(x *Package) []*Package { return x.runClosure() })
		p._runClosureDone = true
	}
	return p._runClosure
}

func (p *Package) iterAllRuntimeDepends() []*Package {
	var result []*Package
	for _, x := range p.runClosure() {
		if x.Buildable() {
			result = append(result, x)
		}
	}
	return result
}
